pkg/controller/certificatesigningrequest/ipsec_csr_signing_controller.go

```diff
@@ -0,0 +1,470 @@
+			},
+			Type: corev1.SecretTypeTLS,
+			Data: map[string][]byte{
+				corev1.TLSCertKey:       caBytes,
+				corev1.TLSPrivateKeyKey: caKeyBytes,
+			},
+		}
+		caSecret, err = c.client.CoreV1().Secrets(env.GetAntreaNamespace()).Create(context.TODO(), caSecret, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+		klog.Info("Created Secret for self-signed IPsec root CA")
+	}
+	caCertificate, err := certutil.ParseCertsPEM(caSecret.Data[corev1.TLSCertKey])
+	if err != nil {
+		return err
+	}
+	if len(caCertificate) == 0 {
+		return fmt.Errorf("CA certificate is empty")
+	}
+	privateKey, err := keyutil.ParsePrivateKeyPEM(caSecret.Data[corev1.TLSPrivateKeyKey])
+	if err != nil {
+		return err
+	}
+	priv, ok := privateKey.(crypto.Signer)
+	if !ok {
+		return fmt.Errorf("error reading CA: key did not implement crypto.Signer")
+	}
+	ca := &certificateAuthority{
+		RawCert:     caSecret.Data[corev1.TLSCertKey],
+		RawKey:      caSecret.Data[corev1.TLSPrivateKeyKey],
+		Certificate: caCertificate[0],
+		PrivateKey:  priv,
+	}
+	c.certificateAuthority.Store(ca)
+	desiredConfigMapData := map[string]string{
+		rootCACertKey: string(caSecret.Data[corev1.TLSCertKey]),
+	}
+	caConfigMap, err := c.configMapLister.ConfigMaps(env.GetAntreaNamespace()).Get(ipsecRootCAName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		caConfigMap = &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      ipsecRootCAName,
+				Namespace: env.GetAntreaNamespace(),
+			},
+			Data: desiredConfigMapData,
+		}
+		caConfigMap, err = c.client.CoreV1().ConfigMaps(env.GetAntreaNamespace()).Create(context.TODO(), caConfigMap, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+		klog.InfoS("Created ConfigMap for self-signed IPsec root CA")
+	}
+	if !reflect.DeepEqual(desiredConfigMapData, caConfigMap.Data) {
+		toUpdate := caConfigMap.DeepCopy()
+		toUpdate.Data = desiredConfigMapData
+		_, err = c.client.CoreV1().ConfigMaps(env.GetAntreaNamespace()).Update(context.TODO(), toUpdate, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *IPsecCSRSigningController) csrWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// watchSecretChanges uses watch API directly to watch for Secret changes.
+// Antrea Controller should not have List permission for Secrets.
+func (c *IPsecCSRSigningController) watchSecretChanges(endCh <-chan struct{}) error {
+	watcher, err := c.client.CoreV1().Secrets(env.GetAntreaNamespace()).Watch(context.TODO(), metav1.SingleObject(metav1.ObjectMeta{
+		Namespace: env.GetAntreaNamespace(),
+		Name:      ipsecRootCAName,
+	}))
+	if err != nil {
+		return fmt.Errorf("failed to create Secret watcher: %v", err)
+	}
+	// re-queue in case of missing events before watcher starts.
+	c.fixturesQueue.Add(workerItemKey)
+	ch := watcher.ResultChan()
+	defer watcher.Stop()
+	for {
+		select {
+		case _, ok := <-ch:
+			if !ok {
+				return fmt.Errorf("event channel closed")
```

Comment:
A watch will only last few hours at most, returning error here would cause error logs generated periodically?

Reply:
I am not aware of this. changed to not return an error. thanks!